kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -284,6 +372,333 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
+  get() {
+    return buildString {
+      val classifierString = when (val cl = classifier) {
+        is KmClassifier.Class -> createClassName(cl.name)
+        is TypeAlias -> createClassName(cl.name)
+        is TypeParameter -> arguments[cl.id].type?.canonicalName ?: "TypeVar(${cl.id})"
+      }
+      append(classifierString)
+
+      val args = arguments.joinToString(", ") {
+        "${it.variance?.name} ${it.type?.canonicalName ?: "*"}".trim()
+      }
+
+      if (args.isNotBlank()) {
+        append('<')
+        append(args)
+        append('>')
+      }
+
+      // TODO not sure if we care about expressing the other type information here
+    }
+  }
+
+/**
+ * Creates a canonical class name as represented in Metadata's [kotlinx.metadata.ClassName], where
+ * package names in this name are separated by '/' and class names are separated by '.'.
+ *
+ * For example: `"org/foo/bar/Baz.Nested"`.
+ *
+ * Local classes are prefixed with ".", but for KotlinPoetMetadataSpecs' use case we don't deal
+ * with those.
+ */
+private fun createClassName(kotlinMetadataName: String): String {
+  require(!kotlinMetadataName.isLocal) {
+    "Local/anonymous classes are not supported!"
+  }
+  // Top-level: package/of/class/MyClass
+  // Nested A:  package/of/class/MyClass.NestedClass
+  val simpleName = kotlinMetadataName.substringAfterLast(
+      '/', // Drop the package name, e.g. "package/of/class/"
+      '.' // Drop any enclosing classes, e.g. "MyClass."
+  )
+  val packageName = kotlinMetadataName.substringBeforeLast("/", missingDelimiterValue = "")
+  val simpleNames = kotlinMetadataName.removeSuffix(simpleName)
+      .removeSuffix(".") // Trailing "." if any
+      .removePrefix(packageName)
+      .removePrefix("/")
+      .let {
+        if (it.isNotEmpty()) {
+          it.split(".")
+        } else {
+          // Don't split, otherwise we end up with an empty string as the first element!
+          emptyList()
+        }
+      }
+      .plus(simpleName)
+
+  return "${packageName.replace("/", ".")}.${simpleNames.joinToString(".")}"
+}
+
+@Suppress("SameParameterValue")
+private fun String.substringAfterLast(vararg delimiters: Char): String {
+  val index = lastIndexOfAny(delimiters)
+  return if (index == -1) this else substring(index + 1, length)
+}
+
+private fun defaultPrimitiveValue(type: Type): Any? =
+    if (type is Class<*> && type.isPrimitive) {
+      when (type) {
+        Boolean::class.java -> false
+        Char::class.java -> 0.toChar()
+        Byte::class.java -> 0.toByte()
+        Short::class.java -> 0.toShort()
+        Int::class.java -> 0
+        Float::class.java -> 0f
+        Long::class.java -> 0L
+        Double::class.java -> 0.0
+        Void.TYPE -> throw IllegalStateException("Parameter with void type is illegal")
+        else -> throw UnsupportedOperationException("Unknown primitive: $type")
+      }
+    } else null
+
+private fun Class<*>.toKmClass(throwOnNotClass: Boolean): KmClass? {
```

Comment:
this wants to be 2 different symbols:

```
  val Class<*>.header: KotlinClassHeader?
  val KotlinClassHeader.kmClass KmClass?
```

That way callers can use `?:` rather than the awkward `throwOnNotClass` thing

Reply:
8b42ed1